server/src/main/java/org/opensearch/gateway/remote/IndexCreationPreIndexMetadataUploadListener.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.gateway.remote;
+
+import org.opensearch.cluster.metadata.IndexMetadata;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+
+/**
+ * Hook for running code that needs to be executed before the upload of index metadata during index creation or
+ * after enabling the remote cluster statement for the first time. The listener is intended to be run in parallel and
+ * async with the index metadata upload.
+ *
+ * @opensearch.internal
+ */
+public interface IndexCreationPreIndexMetadataUploadListener {
+
+    /**
+     * This returns the additional count that needs to be added in the latch present in {@link RemoteClusterStateService}
+     * which is used to achieve parallelism and async nature of uploads for index metadata upload. The same latch is used
+     * for running pre index metadata upload listener.
+     *
+     * @param newIndexMetadataList list of index metadata of new indexes (or first time index metadata upload).
+     * @return latch count to be added by the caller.
+     */
+    int latchCount(List<IndexMetadata> newIndexMetadataList);
+
```

Comment:
Interface looks messier, lets now start with these abstractions

Reply:
there are 2 places where this interface is hooking with the index metadata upload flow. It is bit tricky to have async run also ensuring that it is completed before returning from the upload index metadata method. I can discuss with you on we can simplify this further.